ipa-core/src/protocol/ipa_prf/malicious_security/lagrange.rs

```diff
@@ -0,0 +1,282 @@
+        /// where `c_0` to `c_N` are stored in `polynomial`
+        fn eval<M>(&self, x_output: &GenericArray<F, M>) -> GenericArray<F, M>
+        where
+            M: ArrayLength,
+        {
+            // evaluate polynomial p at evaluation_points and random point using monomial base
+            let mut y_values = GenericArray::generate(|_| F::ZERO);
+            for (x, y) in x_output.iter().zip(y_values.iter_mut()) {
+                // monomial base, i.e. `x^k`
+                let mut base = F::ONE;
+                // evaluate p via `sum_k coefficient_k * x^k`
+                for coefficient in &self.coefficients {
+                    *y += *coefficient * base;
+                    base *= *x;
+                }
+            }
+            y_values
+        }
+    }
+
+    impl<F, N> From<MonomialFormPolynomial<F, N>> for Polynomial<F, N>
+    where
+        F: Field + TryFrom<u128>,
+        <F as TryFrom<u128>>::Error: Debug,
+        N: ArrayLength,
+    {
+        fn from(value: MonomialFormPolynomial<F, N>) -> Self {
+            let canonical_points: GenericArray<F, N> =
+                GenericArray::generate(|i| F::try_from(u128::try_from(i).unwrap()).unwrap());
+            Polynomial {
+                y_coordinates: value.eval(&canonical_points),
+            }
+        }
+    }
+
+    fn lagrange_single_output_point_using_new(
+        output_point: TestField,
+        input_points: [TestField; 32],
+    ) {
+        let polynomial_monomial_form = MonomialFormPolynomial {
+            coefficients: GenericArray::<TestField, U32>::from_array(input_points),
+        };
+        let output_expected = polynomial_monomial_form.eval(
+            &GenericArray::<TestField, U1>::from_array([output_point; 1]),
+        );
+        let polynomial = Polynomial::from(polynomial_monomial_form.clone());
+        let denominator = CanonicalLagrangeDenominator::<TestField, U32>::new();
+        // generate table using new
+        let lagrange_table = LagrangeTable::<TestField, U32, U1>::new(denominator, &output_point);
+        let output = lagrange_table.eval(&polynomial);
+        assert_eq!(output, output_expected);
+    }
+
+    proptest! {
+    #[test]
+    fn proptest_lagrange_single_output_point_using_new(output_point: TestField, input_points in prop::array::uniform32(any::<TestField>())){
+        lagrange_single_output_point_using_new(output_point,input_points);
+    }
+    }
+
+    fn lagrange_canonical_using_from(input_points: [TestField; 8]) {
+        let polynomial_monomial_form = MonomialFormPolynomial {
+            coefficients: GenericArray::<TestField, U8>::from_array(input_points),
+        };
+        // the canonical x coordinates are 0..15, the outputs use coordinates 8..15:
+        let x_coordinates_output = GenericArray::<_, U7>::generate(|i| {
+            TestField::try_from(u128::try_from(i).unwrap() + 8).unwrap()
+        });
```

Comment:
Honestly, I find the generics here(e.g. U7, U8) really confusing. I wish there was some better way.